### PR TITLE
Add union-typed values to ztests

### PR DESCRIPTION
**Pull request by philrz**
## What's Changing

This adds union-typed values to ztests that didn't have them before.

## Why

This is a complement to the many bugs I've opened lately about functions/operators/etc. that had problems with union-typed values. For the many I tested that _weren't_ buggy, it seems worthwhile to have the union test coverage so we can avoid regressions.

## Details

At a couple spots while I was making these changes I did need to do some close reading of the docs, e.g., confirming where dropping the union type info on output or returned values is _expected_ behavior, as I was nervous I might be putting in tests that actually lock down bad behavior. I _think_ I got 'em all right, but thorough review is appreciated!

### CODE REVIEW

File: runtime/ztests/expr/function/compare.yaml [hunk 1]
```diff
@@ -29,8 +33,16 @@ vector: true
 
 input: |
   {a:null,b:2}
+  {a:null::(int64|null),b:2}
+  {a:null::(int64|null),b:2::(int64|null)}
```
Comment (philrz) [inline runtime/ztests/expr/function/compare.yaml:37]: Note that a fix for #6924 should probably add some more test values here that have a union type like `(bool|null)` for the `nullsMax` argument, as that's not working yet.

**Review by nwt — APPROVED** [review]

File: runtime/ztests/expr/function/grep.yaml [hunk 2]
```diff
@@ -16,6 +16,8 @@ input: |
   {pattern:"z",input:{a:{b:"c"}}}
   {pattern:1,input:""}
   {pattern:null,input:"a"}
+  {pattern:"a",input:"hello"::(string|null)}
+  {pattern:"a",input:null::(string|null)}
```
Comment (philrz) [inline runtime/ztests/expr/function/grep.yaml:20]: Note that a fix for #6925 should probably add some more test values here that have a union type for the `pattern` argument, as that's not working yet.

File: runtime/ztests/expr/function/nameof.yaml [hunk 3]
```diff
@@ -18,6 +18,8 @@ input: |
   {x:"foo",y:1,z:2}::bar2
   <{x:string,y:int64,z:int64}>
   null
+  1::(int64|null)
+  null::(int64|null)
```
Comment (philrz) [inline runtime/ztests/expr/function/nameof.yaml:22]: Note that a fix for #6927 should probably add some more test values here that have a named type attached to a union type, as that's not working yet.

File: runtime/ztests/op/assert.yaml [hunk 4]
```diff
@@ -6,8 +6,12 @@ input: |
   {a:1}
   {a:2}
   1
+  {a:1::(int64|null)}
+  {a:2::(int64|null)}
```
Comment (philrz) [inline runtime/ztests/op/assert.yaml:10]: Note that the test passes with these values as record fields, but as shown in #6907 and fixed in #6926, it would have failed with top-level values. Tricky!
